### DISCUSSION TITLE
Loading component for all pages

### DISCUSSION THREAD

**Discussion by JorensM** [vercel/next.js#61324] [Ideas]
### Goals

1. Make it easier to create a loading component when you want a single one for the entire app
2.
3.

### Non-Goals

1.
2.
3.

### Background

I'm just trying out Next.JS and found it strange that you must define a loading.tsx component for each page separately for it to show. I'd like for it to be possible show a single loading component between every single page in the app without the need to create a separate `loading.tsx` file for each page

### Proposal

Not sure how to implement this but I'm interested in contributing to this.

**Comment by 808vita** [comment]
**If you are using `app router` use this [doc link](https://nextjs.org/docs/app/building-your-application/upgrading/app-router-migration#step-2-creating-a-root-layout)**
Add your component into layout.tsx . pasting snippet below for convenience. 

Create a new app/layout.tsx file inside the app directory. This is a [root layout](https://nextjs.org/docs/app/building-your-application/routing/pages-and-layouts#root-layout-required) that will apply to all routes inside app.

`app/layout.tsx`

```
TypeScript

export default function RootLayout({
  // Layouts must accept a children prop.
  // This will be populated with nested layouts or pages
  children,
}: {
  children: React.ReactNode
}) {
  return (
    <html lang="en">
      <body>{children}</body>
    </html>
  )
}
```
The app directory must include a root layout.
The root layout must define <html>, and <body> tags since Next.js does not automatically create them
The root layout replaces the pages/_app.tsx and pages/_document.tsx files.
.js, .jsx, or .tsx extensions can be used for layout files.

___

**If you are using `pages router` use this [doc link](https://nextjs.org/docs/pages/building-your-application/routing/custom-app)**
Add your component into `_app.tsx` . pasting snippet below for convenience. 

[Usage](https://nextjs.org/docs/pages/building-your-application/routing/custom-app#usage)
To override the default App, create the file pages/_app as shown below:

`pages/_app.tsx`

```
TypeScript

import type { AppProps } from 'next/app'
 
export default function MyApp({ Component, pageProps }: AppProps) {
  return <Component {...pageProps} />
}
```

Good to know
If your app is running and you added a custom App, you'll need to restart the development server. Only required if pages/_app.js didn't exist before.

**Reply by JorensM** [reply]
You must've misunderstood my proposal. I already have a root layout. I'm talking about the [Loading component](https://nextjs.org/docs/app/building-your-application/routing/loading-ui-and-streaming) that you can define in a `loading.tsx` file

**Reply by damianlluch** [reply]
This is valid for NextJS 14? I have created the _app.tsx inside the app folder and in the root of the project and I can't get it to run.

**Comment by alenvuletic** [comment]
+1

**Comment by aprilmintacpineda** [comment]
Yes please, I've experimented on this and found the following:

- I can add `app/error.tsx` and if I throw errors on any pages, this error page will be used when there's no `error.tsx` on that page.
- I can add `app/loading.tsx` which didn't seem to take effect on the individual pages, when there's no `loading.tsx` on the page itself, nothing will be displayed as loading indicator on the client, and the client has to wait for the new page to finish rendering before the route changes, unlike when there's a `loading.tsx` on the page, the route changes immediately and renders the loading indicator.

**Reply by sadekomar** [reply]
I'm having the same issue. Have you found a way around it yet?

**Comment by JuanGarcia0323** [comment]
The closest thing that I found is defining a component Loading.tsx, importing it inside my folder and exporting it 

![image](https://github.com/user-attachments/assets/a474029c-784f-4c62-86a4-d372fc13137c)

`
import Loading from "@/components/ui/loading";
export default Loading;
`

**Comment by alinchiuaru** [comment]
Just stumbled into this, the way I handled this was to wrap the children from the root layout with a Suspense and have the fallback be the loading component.

```
// app/layout.ts
export default function RootLayout({
  children,
}: {
  children: React.ReactNode
}) {
  return (
    <html lang="en">
      <body>
        {/* Layout UI */}
        {/* Place children where you want to render a page or nested layout */}
        <Suspense fallback={<Loading />}>
            {children}
        </Suspense>
      </body>
    </html>
  )
}
```
